Add an How To's for HTTP2 and SSL (#453)
meilisearch/documentation: resources/howtos/http2_ssl.md

---
permalink: /http2-ssl
---

> # How to use HTTP2 and SSL with MeiliSearch

Reviewer: ```suggestion
# How to use HTTP/2 and SSL with MeiliSearch
```

> For those who would like to use HTTP2 we have to warn you that this is only possible if you have launched your server with SSL.

Reviewer 2: ```suggestion
For those who would like to use HTTP2 we have to warn you that this is **only possible if you have launched your server with SSL**.
```

> For those who would like to use HTTP2 we have to warn you that this is **only possible if you have launched your server with SSL**.

Reviewer: ```suggestion
For those willing to use HTTP/2, please be aware that it is **only possible if your server is configured with SSL certificate**.
```

We will therefore see here how to launch a MeiliSearch server with SSL. In this little tutorial I will do it locally, but be aware that you can very well do the same thing on a server.

> To start I need the binary of MeiliSearch, we could also use docker, it would be necessary to pass it the parameters in environment variables and the certificates via a volume.

Reviewer: ```suggestion
First of all, you need the binary of MeiliSearch, or you can also use docker. In the latter case, it is necessary to pass the parameters using environment variables and the SSL certificates via a volume.
```

> To start I need the binary of MeiliSearch, we could also use docker, it would be necessary to pass him the parameters in variable of environment aison that the certificates via a volume.

Reviewer 2: ```suggestion
To start I need the binary of MeiliSearch, we could also use docker, it would be necessary to pass it the parameters in environment variables aison that the certificates via a volume.
```
`aison` ?
I don't understand the end of the sentence

Reviewer 2: `aison` ?
I don't understand the end of the sentence

> To start I need the binary of MeiliSearch, we could also use docker, it would be necessary to pass it the parameters in environment variables aison that the certificates via a volume.

Author: ```suggestion
To start I need the binary of MeiliSearch, we could also use docker, it would be necessary to pass it the parameters in environment variables and the certificates via a volume.
```

> We will also need a tool to generate a certificate. In this example I use [mkcert](https://github.com/FiloSottile/mkcert). We could also use certbot if we were on a server or directly signed certificates by a certificate authority.

Reviewer: ```suggestion
A tool to generate SSL certificates is also required. In this How To, you will use [mkcert](https://github.com/FiloSottile/mkcert). However, if on a remote server, you can also use certbot or certificates signed by a Certificate Authority.
```

> Finally, I use curl to make my requests because it's easy to specify that we want to make HTTP2 requests with the `--http2` option.

Reviewer: ```suggestion
Then, use curl to make requests. It is a simple way to specify that you want to make HTTP/2 requests by using the `--http2` option.
```

> Finally, here I would use curl to make my requests because it's easy to specify that we want to make HTTP2 requests with the `--http2` option.

Reviewer 2: ```suggestion
Finally, I use curl to make my requests because it's easy to specify that we want to make HTTP2 requests with the `--http2` option.
```

> ## Test with HTTP2 without SSL

Reviewer: ```suggestion
## Test with HTTP/2 without SSL
```

> We start by running the binary.

Reviewer: ```suggestion
Start by running the binary.
```

```
./meilisearch
```

> And then test a request.

Reviewer: ```suggestion
And then, test a request.
```

```
curl -kvs --http2 --request GET 'http://localhost:7700/indexes'
```

> The answer from the server is:

Reviewer: ```suggestion
You will get the following answer from the server:
```

```
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 7700 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 7700 (#0)
> GET /indexes HTTP/1.1
> Host: localhost:7700
> User-Agent: curl/7.64.1
> Accept: */*
> Connection: Upgrade, HTTP2-Settings
> Upgrade: h2c
> HTTP2-Settings: AAMAAABkAARAAAAAAAIAAAAA
>
< HTTP/1.1 200 OK
< content-length: 2
< content-type: application/json
< date: Fri, 17 Jul 2020 11:01:02 GMT
<
* Connection #0 to host localhost left intact
[]* Closing connection 0
```

> we can see with the line `> Connection: Upgrade, HTTP2-Settings` that the server tries to upgrade to HTTP2 but fails.

Reviewer: ```suggestion
You can see on line `> Connection: Upgrade, HTTP2-Settings` that the server tries to upgrade to HTTP/2, but is unsuccessful. 
```

> You can see on line `> Connection: Upgrade, HTTP2-Settings` that the server tries to upgrade to HTTP/2, but is unsuccessful. 

Reviewer 2: ```suggestion
You can see on line `> Connection: Upgrade, HTTP2-Settings` that the server tries to upgrade to HTTP/2, but is unsuccessful.
```

This is why the styling check failes!

> The answer `< HTTP/1.1 200 OK` indicates that we stayed in HTTP1.

Reviewer: ```suggestion
The answer `< HTTP/1.1 200 OK` indicates that the server still uses HTTP/1.
```

> ## Test with HTTP2 with SSL

Reviewer: ```suggestion
## Try to use HTTP/2 with SSL
```

> This time we start by generating the SSL certificates. This one generates two files `127.0.0.1.pem` and `127.0.0.1-key.pem`.

Reviewer: ```suggestion
This time, start by generating the SSL certificates. Two files are generated: `127.0.0.1.pem` and `127.0.0.1-key.pem`.
```

```
mkcert '127.0.0.1'
```

> Then we use the certificate and the key to launch MeiliSearch with SSL.

Reviewer: ```suggestion
Then, use the certificate and the key to configure MeiliSearch with SSL.
```

```
./meilisearch --ssl-cert-path ./127.0.0.1.pem --ssl-key-path ./127.0.0.1-key.pem
```

Next, make the same request as above but change `http://` to `https://`.

```
curl -kvs --http2 --request GET 'https://localhost:7700/indexes'
```

> The answer from the server is:

Reviewer: ```suggestion
You will get the following answer from the server:
```

* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: O=mkcert development certificate; OU=quentindequelen@s-iMac (Quentin de Quelen)
*  start date: Jun  1 00:00:00 2019 GMT
*  expire date: Jul 17 10:38:53 2030 GMT
*  issuer: O=mkcert development CA; OU=quentindequelen@s-iMac (Quentin de Quelen); CN=mkcert quentindequelen@s-iMac (Quentin de Quelen)
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7ff601009200)
> GET /indexes HTTP/2
> Host: localhost:7700
> User-Agent: curl/7.64.1
> Accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS == 4294967295)!
< HTTP/2 200
< content-length: 2
< content-type: application/json
< date: Fri, 17 Jul 2020 11:06:27 GMT
<
* Connection #0 to host localhost left intact
[]* Closing connection 0
```

> We can see here that the server has successfully switched from HTTP1 to HTTP2.

Reviewer: ```suggestion
You can see that the server now supports HTTP/2.
```

```
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
```

> The server responds well in HTTP2.

Reviewer: ```suggestion
The server successfully receives HTTP/2 requests.
```
